js/misc/util.js

```diff
@@ -403,6 +404,240 @@ function findIndex(array, callback) {
     return -1;
 }
 
+/**
+ * find:
+ * @array (array): Array to be iterated.
+ * @callback (function): The function to call on every iteration,
+ * should return a boolean value.
+ *
+ * Returns (any): Returns the matched element, else null.
+ */
+function find(arr, callback) {
+    for (let i = 0, len = arr.length; i < len; i++) {
+        if (callback(arr[i], i, arr)) {
+            return arr[i];
+        }
+    }
+    return null;
+};
+
+/**
+ * each:
+ * @array (array|object): Array or object to be iterated.
+ * @callback (function): The function to call on every iteration.
+ *
+ * Iteratee functions may exit iteration early by explicitly returning false.
+ */
+function each(obj, callback) {
+    if (Array.isArray(obj)) {
+        for (let i = 0, len = obj.length; i < len; i++) {
+            let returnValue = callback(obj[i], i);
+            if (returnValue === false) {
+                return;
+            }
+        }
+    } else {
+        let keys = Object.keys(obj);
+        for (let i = 0, len = obj.length; i < len; i++) {
+            let key = keys[i];
+            callback(obj[key], key);
+        }
+    }
+};
+
+/**
+ * filter:
+ * @array (array): Array to be iterated.
+ * @callback (function): The function to call on every iteration.
+ *
+ * Returns (array): Returns the new filtered array.
+ */
+function filter(arr, callback) {
```

Comment:
Aren't `filter`, `each` and `find` already built into cjs arrays? No need to duplicate them here.

Reply:
They are, but on Spidermonkey there is a performance gain to using for loop alternatives instead. This is a less of an issue on V8. I ran these tests in the Firefox 52 ESR, which is the closest to our environment.

https://jsperf.com/aarontgrogg-array-filter-vs-for-loop
https://jsperf.com/loops/33
http://jsben.ch/ABI1C
http://jsben.ch/aY0kG

Reply:
Those tests compare filter vs raw code, but that is not the case here because at the end what slows it down is calling the callback function I guess. I don't know about the implementation but sure is very similar (?). I'm going to test it on Cinnamon anyway, just because I'm curious :)

Reply:
Cool. The jsben.ch tests are ones I set up - here's one similar to the findIndex one, but with filter: http://jsben.ch/umjQN

Reply:
My results conclude that you are right, the loop implementation is faster BUT, the difference in my tests for arrays of 10M elements is a couple of ms (and it even becomes zero the more you use it). For arrays of 100 elements (if you go crazy) the difference is non existent. I think native methods should be used for consistency, or preferably normal loops which is perfectly fine, but defining those seems like not ok at all. It's just my opinion :)

[Here are my tests](https://drive.google.com/open?id=1cPWm7nbT57zC4d-glCi1XUGRau_WciLC) (it is a cinn extension).

Reply:
Thanks for looking into this more. While I agree this won't make a dramatic difference in a lot of cases, I think these methods should stay in the PR. Take a function that iterates multiple things to get its return value and gets called relatively frequently for instance, for invocation the latency can add up - especially when we want Cinnamon to be as responsive.

These are there to provide an option, not to replace every instance of filter, findIndex, etc - as you point out the gains diminish with smaller arrays, but the gains are still there and consistent.

Here's the results I got from test-extension:

Cjs-Message: 13:45:25.762: JS LOG: [LookingGlass/info] Array.prototype.filter: 20.6ms
Cjs-Message: 13:45:25.763: JS LOG: [LookingGlass/info] Custom filter:              11.7ms

If this were a function that was meant to assist with an animation, it could mean the difference between something being jerky or smooth, since CInnamon needs to spend 16ms or less for each frame.

Reply:
I understand your point, but in those cases why not to use loops? I presume they are even faster than these functions.

PS: those results are from 1M array items and appear like the first time or so you use the functions, then they are like the same time. In an animation there would be no difference for example, assuming you have an array of <1M items.

Reply:
I think you're seeing that because of Spidermonkey's JIT optimization- if it doesn't get GC'd its still cached as byte code, but we shouldn't rely on that. 

This allows the performance benefit of for loops, but the readability of explicitly named functions like forEach. The each function I added in particular can iterate both arrays and objects, which can be convenient. For objects it uses Object.keys and a for loop, which is a little faster than for...in.

Reply:
Sorry @jaszhix I still don't get where is exactly that critical place in this PR where you have a large array and save more than 1 ms. I think that such micro-optimisations belong to cjs, not cinnamon. But idk 😅

Reply:
Yes, of course. But why not implement them in cjs instead? If we can make native methods faster why can't we improve them instead of creating faster duplicates so that cinnamon can benefit from that in every single place? It seems like a better solution to me, but maybe there is a reason not to do that I can't see.

Reply:
Mainly because I'd prefer to keep CJS even as possible with GJS, otherwise back-porting patches gets harder. I could upstream it, but not sure they would want them. If they do, great, but then we can't add the optimizations on our schedule.

Reply:
Oh I understand. Maybe some day! Sorry for bothering with this 😅